### PR TITLE
Computer and code setup

### PR DESCRIPTION
**Pull request by yakutovicha**
Here is my work on computer and code setup.

In brief what I've implemented:

1) In `CodeDropdown widget` I added a `refresh` button and `setup another` link.

2) `setup another` link will bring the user to code setup notebook, which is based on `AiiDACodeSetup` widget. The notebook also accepts initial values of the parameters through URL:

```
from aiidalab_widgets_base.codes import AiiDACodeSetup, extract_aiidacodesetup_arguments

parsed_url = urlparse.parse_qs(urlparse.urlsplit(jupyter_notebook_url).query)
args = extract_aiidacodesetup_arguments(parsed_url)

code = AiiDACodeSetup(**args)
display(code)
```

3)  `AiiDACodeSetup` widget also contains `ComputerDropdown` widget that allows to select computers that are already setup. 

4) The `ComputerDropdown` widget also contains a link to setup a new computer.

5) The computer setup notebook is based on  `SshComputerSetup` and `AiidaComputerSetup` widgets. They are used to configure password-free access to remote machine and to configure the computer with AiiDA. Both widgets accept initial values for the parameters (in same way as  `AiiDACodeSetup` does).

### CODE REVIEW

**Comment by yakutovicha** [comment]
@ltalirz would you like to try it out, or should I just go on with squaching/merging?

**Comment by ltalirz** [comment]
I'll have a look

**Comment by yakutovicha** [comment]
@CasperWA if you have time, please have a look also

File: aiidalab_widgets_base/codes.py [hunk 1]
```diff
@@ -2,7 +2,30 @@
...
 from __future__ import absolute_import
 import ipywidgets as ipw
+
+from subprocess import check_output
 from IPython.display import clear_output
+from aiida import load_dbenv, is_dbenv_loaded
+from aiida.backends import settings
+if not is_dbenv_loaded():
+    load_dbenv(profile=settings.AIIDADB_PROFILE)
+from aiida.orm import Code, Computer
```
Comment (CasperWA) [inline aiidalab_widgets_base/codes.py]: `Computer` is only used for `_get_codes`. The import could be moved there.

File: aiidalab_widgets_base/codes.py [hunk 2]
```diff
@@ -62,7 +83,7 @@ def _get_codes(self, input_plugin):
         codes = {"{}@{}".format(r[1].label, r[0].name): r[1] for r in results}
         return codes
 
-    def refresh(self):
+    def refresh(self, b=None):
```
Comment (CasperWA) [inline aiidalab_widgets_base/codes.py]: Why `b`?

File: aiidalab_widgets_base/computers.py [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+    return valid_arguments(arguments, VALID_SSH_COMPUTER_SETUP_ARGUMETNS)
+
+def extract_aiidacomputer_arguments(arguments):
+    return valid_arguments(arguments, VALID_AIIDA_COMPUTER_SETUP_ARGUMETNS)
+
+class SshComputerSetup(ipw.VBox):
+    def __init__(self, **kwargs):
+        style = {"description_width":"200px"}
+        computer_image = ipw.HTML('<img width="200px" src="./miscellaneous/images/computer.png">')
+        self._inp_username = ipw.Text(description="Ssh username:",
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: "SSH"

Comment (CasperWA) [reply aiidalab_widgets_base/computers.py]: Same below

File: aiidalab_widgets_base/computers.py [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+                                        self._proxy_user_password_box],
+                                 layout = {'visibility':'hidden'})
+
+        self._btn_setup_ssh = ipw.Button(description="Setup ssh")
+        self._btn_setup_ssh.on_click(self.on_setup_ssh)
+        self._setup_ssh_out = ipw.Output()
+
+        # Check if some settings were already provided
+        self._predefine_settings(**kwargs)
+        children = [ipw.HBox([computer_image, ipw.VBox([self._inp_computer_hostname,
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: For me, the `computer_ssh_box` and `_proxy_ssh_box` cannot be fully shown in the width, and is scrollable.

I think this can be mediated by changing margins and pixel-widths

File: aiidalab_widgets_base/computers.py [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+            self._proxy_ssh_box.layout.visibility = 'hidden'
+            self._use_diff_proxy_username.value = False
+
+    def on_use_diff_proxy_username_change(self, b):
+        if self._use_diff_proxy_username.value:
+            self._proxy_user_password_box.layout.visibility = 'visible'
+        else:
+            self._proxy_user_password_box.layout.visibility = 'hidden'
+    
+    def on_computer_hostname_change(self, b):
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: This is nice, but is run every single time a character is typed in or deleted.

File: aiidalab_widgets_base/computers.py [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+    def __init__(self, **kwargs):
+        from aiida.transport import Transport
+        from aiida.scheduler import Scheduler
+        style = {"description_width":"200px"}
+
+        # list of widgets to be displayed
+        self._btn_setup_comp = ipw.Button(description="setup computer")
+        self._btn_setup_comp.on_click(self._on_setup_computer)
+        self._inp_computer_name = ipw.Text(value='',
+                                           placeholder='Will be only used within AiiDA',
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: "Will only be used within AiiDA"

File: aiidalab_widgets_base/computers.py [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+                                          description="Workdir:",
+                                          layout=ipw.Layout(width="500px"),
+                                          style=style)
+        self._computer_mpirun_cmd = ipw.Text(value='mpirun -n {tot_num_mpiprocs}',
+                                             description="Mpirun command:",
+                                             layout=ipw.Layout(width="500px"),
+                                             style=style)
+        self._computer_ncpus = ipw.IntText(value=12,
+                                           step=1,
+                                           description='Number of CPU(s) per node:',
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: No parenthesis needed

File: aiidalab_widgets_base/computers.py [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+                                       style=style)
+        self._prepend_text = ipw.Textarea(placeholder='Text to prepend to each command execution',
+                                          description='Prepend text:',
+                                          layout=ipw.Layout(width="400px")
+                                         )
+        self._append_text = ipw.Textarea(placeholder='Text to append to each command execution',
+                                         description='Append text:',
+                                         layout=ipw.Layout(width="400px")
+                                        )
+        self._btn_test = ipw.Button(description="test computer")
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: "Test computer"

File: aiidalab_widgets_base/computers.py [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+                   ipw.HBox([self._setup_comp_out, self._test_out]),
+                  ]
+        super(AiidaComputerSetup, self).__init__(children, **kwargs)
+
+    def _predefine_settings(self, **kwargs):
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+            else:
+                raise AttributeError("'{}' object has no attirubte '{}'".format(self, key))
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: attribute

File: aiidalab_widgets_base/computers.py [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+        """ Dropdown for Codes for one input plugin.
+
+        :param text: Text to display before dropdown
+        :type text: str
+        """
+
+        self._dropdown = ipw.Dropdown(options=[], description=text, style={'description_width': 'initial'}, disabled=True)
+        self._btn_refresh = ipw.Button(description="refresh", layout=ipw.Layout(width="70px"))
+
+        self._setup_another = ipw.HTML(value="""<a href=./setup_computer.ipynb target="_blank">setup another</a>""")
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: Setup new computer

File: setup_computer.ipynb [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    "args =  extract_sshcomputersetup_arguments(parsed_url)\n",
+    "sshcomputer = SshComputerSetup(**args)\n",
+    "display(sshcomputer)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 2: Setup&Test AiiDA Computer"
```
Comment (CasperWA) [inline setup_computer.ipynb]: Setup & Test AiiDA Computer

File: aiidalab_widgets_base/codes.py (hunk 1 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: done

File: aiidalab_widgets_base/codes.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: yes, otherwise it does not work

File: aiidalab_widgets_base/codes.py [hunk 12]
```diff
@@ -18,23 +41,21 @@ def __init__(self, input_plugin, text='Select code:', **kwargs):
         self.input_plugin = input_plugin
         self.codes = {}
 
-        self.label = ipw.Label(value=text)
-        self.dropdown = ipw.Dropdown(options=[], disabled=True)
+        self.dropdown = ipw.Dropdown(description=text, disabled=True)
+        self._btn_refresh = ipw.Button(description="refresh", layout=ipw.Layout(width="70px"))
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: done

File: aiidalab_widgets_base/codes.py [hunk 13]
```diff
@@ -18,23 +41,21 @@ def __init__(self, input_plugin, text='Select code:', **kwargs):
...
+        self._btn_refresh.on_click(self.refresh)
+        self._setup_another = ipw.HTML(value="""<a href=./setup_code.ipynb target="_blank">setup another</a>""")
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: done

File: aiidalab_widgets_base/codes.py [hunk 14]
```diff
@@ -83,3 +104,158 @@ def selected_code(self):
...
+
+        self._prepend_text = ipw.Textarea(placeholder='Text to prepend to each command execution',
+                                          description='Prepend text:',
+                                          layout=ipw.Layout(width="400px"))
+
+        self._append_text = ipw.Textarea(placeholder='Text to append to each command execution',
+                                         description='Append text:',
+                                         layout=ipw.Layout(width="400px"))
+
+        self._btn_setup_code = ipw.Button(description="setup code")
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: done

File: aiidalab_widgets_base/codes.py [hunk 15]
```diff
@@ -83,3 +104,158 @@ def selected_code(self):
...
+        # Check if some settings were already provided
+        self._predefine_settings(**kwargs)
+        super(AiiDACodeSetup, self).__init__(children, **kwargs)
+
+    def _predefine_settings(self, **kwargs):
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+            else:
+                raise AttributeError("'{}' object has no attirubte '{}'".format(self, key))
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: done

File: aiidalab_widgets_base/computers.py (hunk 3 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: done

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: To best of my knowledge [this](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Events.html#Continuous-updates) is the only way to control the behavior.
I tried `continuous_update=False`, but this does not work properly

File: aiidalab_widgets_base/computers.py (hunk 6 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: done

File: aiidalab_widgets_base/computers.py (hunk 7 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: why? it can also be 1 cpu...

File: aiidalab_widgets_base/computers.py (hunk 9 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: :)

File: aiidalab_widgets_base/computers.py (hunk 10 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: done

File: setup_computer.ipynb (hunk 11 above)
Comment (yakutovicha) [reply setup_computer.ipynb]: done

File: aiidalab_widgets_base/computers.py (hunk 7 above)
Comment (ltalirz) [reply aiidalab_widgets_base/computers.py]: you would still say "number of central processing units"

File: aiidalab_widgets_base/computers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: fixed

File: aiidalab_widgets_base/codes.py [hunk 16]
```diff
@@ -2,7 +2,30 @@
 
 from __future__ import absolute_import
 import ipywidgets as ipw
+
+from subprocess import check_output
 from IPython.display import clear_output
+from aiida import load_dbenv, is_dbenv_loaded
+from aiida.backends import settings
+if not is_dbenv_loaded():
```
Comment (ltalirz) [inline aiidalab_widgets_base/codes.py]: We are intentionally avoiding `load_dbenv` at module import in the other widgets.
Why should we break this here?

Note that the `__init__.py` imports this file.

Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: ok, as we agreed I will replace `load_dbenv` function with [with_dbenv](https://github.com/aiidateam/aiida_core/blob/aaa8c687bd325bedf1d4aeca479d4ca9a2b96eb4/aiida/cmdline/dbenv_lazyloading.py#L38-L56) decorator on `__init__()` function of each class.

File: aiidalab_widgets_base/computers.py (hunk 7 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: Ok

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (CasperWA) [reply aiidalab_widgets_base/computers.py]: Okay. From your link it also seems this only works for Text and Int widgets. I am also not sure exactly what can be done with the `widgets.link` function.

File: aiidalab_widgets_base/codes.py (hunk 16 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/codes.py]: ok, so we change the final decision. This code will be put in the top `__init__.py` file of the module:
```
from aiida import load_dbenv, is_dbenv_loaded
from aiida.backends import settings
if not is_dbenv_loaded():
    load_dbenv(profile=settings.AIIDADB_PROFILE)
```

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (CasperWA) [reply aiidalab_widgets_base/computers.py]: It worked for me when initiating `_inp_computer_hostname` with `contiunous_update=False`. One simply has to stop editing the text field, and then it will be updated.

File: aiidalab_widgets_base/computers.py (hunk 4 above)
Comment (CasperWA) [reply aiidalab_widgets_base/computers.py]: It should be possible to keep the password textbox in the same size as the username textbox, since the `Password` widget inherits from `Text`, but if it's too difficult/fiddly, it is also okay as it is now. It's purely aesthetic in the end.

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: Try to remove the text from the hostname field -> you will still be able to press the `Get from config` button.

**Comment by yakutovicha** [comment]
I think I am done.

@ltalirz and @CasperWA need your final OK to merge.

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (CasperWA) [reply aiidalab_widgets_base/computers.py]: And then immediately you won't be able to.
The only problem is if, when you press the button for that edge-case it takes the value from the text field to be empty or what it originally was before you removed the test?

File: aiidalab_widgets_base/computers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: Well, of course if you know a way to fix that - I am happy to add your fix.

I tried to play with margins, width parameters. This did not result in any success.

To my understanding the only parameter that Password widget takes into account is `width="138px"` which affects the description width only.

**Comment by CasperWA** [comment]
> I think I am done.

You don't want to change or decide on the last thing in my review first?

**Comment by yakutovicha** [comment]
> You don't want to change or decide on the last thing in my review first?

I definitely want

File: aiidalab_widgets_base/computers.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: just to say that after discussion @ltalirz and @CasperWA we decided to completely remove this button, as it's functionality probably not needed.

**Comment by yakutovicha** [comment]
pinging @CasperWA

File: aiidalab_widgets_base/computers.py [hunk 17]
```diff
@@ -39,32 +40,24 @@ def extract_aiidacomputer_arguments(arguments):
     return valid_arguments(arguments, VALID_AIIDA_COMPUTER_SETUP_ARGUMETNS)
 
 class SshComputerSetup(ipw.VBox):
+    setup_counter = Int(0) # Traitlet to inform other widgets about changes
     def __init__(self, **kwargs):
         style = {"description_width":"200px"}
         computer_image = ipw.HTML('<img width="200px" src="./miscellaneous/images/computer.png">')
         self._inp_username = ipw.Text(description="SSH username:",
                                       layout=ipw.Layout(width="350px"),
+                                      contiunous_update=False,
```
Comment (CasperWA) [inline aiidalab_widgets_base/computers.py]: continuous_update

Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: ah, this I actually wanted to delete. Sorry, leftover from my tests

Comment (yakutovicha) [reply aiidalab_widgets_base/computers.py]: fixed

**Comment by ltalirz** [comment]
@yakutovicha Could we make it a rule that new widgets added are documented in the README?

Otherwise we end up with a lot of undocumented stuff and people will won't be aware it exists.

**Comment by yakutovicha** [comment]
@ltalirz Sure, I am going to add a description in the readme in a separate branch (before merging into master and making a new release)
